src/GafferSceneUI/LightTool.cpp

```diff
@@ -855,11 +862,24 @@ class SpotLightHandle : public LightToolHandle
 
 		bool visible() const override
 		{
-			if( !m_coneAngleInspector )
+			if( !m_coneAngleInspector || ( m_handleType == HandleType::Penumbra && !m_penumbraAngleInspector ) )
+			{
+				return false;
+			}
+
+			// Check the inspection results for the current context, which may be set for a non-spot light location
+
+			Inspector::ResultPtr contextConeInspection = m_coneAngleInspector->inspect();
+			Inspector::ResultPtr contextPenumbraInspection = m_penumbraAngleInspector ? m_penumbraAngleInspector->inspect() : nullptr;
+
+			if( !contextConeInspection || ( m_handleType == HandleType::Penumbra && !contextPenumbraInspection ) )
```

Comment:
It feels a bit weird for me that we grab these Inspections to test if they're null, but we never use them for anything ... it makes it unclear why we need to test them.

Are these different from the coneInspection and penumbraInspection values returned from spotLightHandleAngles ... I can't really tell how they differ? If they are the same, then I would find it clearer to call spotLightHandleAngles first, and then test the values we are actually using for null inspections.

Reply:
They are slightly different from `spotLightHandleAngles()`. That function returns the handle angles and inspections for the scene location where the handles will be attached, the last selected location. These inspections are done in the current context, which for this purpose is set in a loop over the selected locations starting at line 1699. If the inspection is null, it means we can't do anything for this scene location (probably because it's not a spot light) so the handle should be hidden. 

This way we only show the handles if the handle for all of the selection is visible.

I did that in order to keep the selection handling in `LightTool` and not make each individual tool have to keep track of it, which still sounds like a good idea to me but I'm open to suggestions.

Perhaps a better comment is enough to clarify what's being tested?

Reply:
Oh, now that you've explained it, that does make sense. I think part of the reason it feels weird is that the loop on line 1699 is visiting all the locations in the selection, but the penumbra logic in visible() is just based on the last selected, so each of the visible() calls ends up evaluating the penumbra of the same location, right? Feels a little bit odd to me, but you've thought about this interface way more than I have. I certainly don't see any reason to hold this up any further, but  maybe a comment wouldn't hurt.

Reply:
> so each of the visible() calls ends up evaluating the penumbra of the same location, right?

Yes, which is not ideal. This is a unique check to spot light penumbras, as least so far, and I'm not sure where would be a better place to put it. Maybe just in the drawing code?

The same caching optimization the drawing code benefits from applies here too, so I don't think it will lead to performance problems. I'm not quite ready to change it without a more thorough audit of the interface overall, so I've added it as a to-do for now.

Maybe after the quad light tool is done I can revisit this overall and have a good picture of what improvements can be made for future maintainability.

For now I've squashed more explanatory comments into https://github.com/GafferHQ/gaffer/pull/5218/commits/49b8945d19d6aecda49d10c3cba482e97b4843cc

Reply:
Looks fine for now.